guides/keras_cv/simsiam-with-kerascv.py

```diff
@@ -0,0 +1,604 @@
+predictor.summary()
+
+
+"""
+## Training
+
+First, we need to initialize our training model, loss, and optimizer.
+"""
+loss = tfsim.losses.SimSiamLoss(projection_type="cosine_distance", name="simsiam")
+
+contrastive_model = tfsim.models.ContrastiveModel(
+    backbone=backbone,
+    projector=projector,
+    predictor=predictor,  # NOTE: simiam requires predictor model.
+    algorithm="simsiam",
+    name="simsiam",
+)
+lr_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=INIT_LR,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+wd_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=WEIGHT_DECAY,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+optimizer = tfa.optimizers.SGDW(
+    learning_rate=lr_decayed_fn, weight_decay=wd_decayed_fn, momentum=0.9
+)
+
+"""
+Next we can compile the model the same way you compile any other Keras model.
+"""
+
+contrastive_model.compile(
+    optimizer=optimizer,
+    loss=loss,
+)
+
+"""
+We track the training using several callbacks.
+
+* **EvalCallback** creates an index at the end of each epoch and provides a proxy for the nearest neighbor matching classification using `binary_accuracy`.
+* **TensordBoard** and **ModelCheckpoint** are provided for tracking the training progress.
+"""
+
+DATA_PATH = Path("./")
+log_dir = DATA_PATH / "models" / "logs" / f"{loss.name}_{time.time()}"
+chkpt_dir = DATA_PATH / "models" / "checkpoints" / f"{loss.name}_{time.time()}"
+
+callbacks = [
+    tfsim.callbacks.EvalCallback(
+        tf.cast(x_query, tf.float32),
+        y_query,
+        tf.cast(x_index, tf.float32),
+        y_index,
+        metrics=["binary_accuracy"],
+        k=1,
+        tb_logdir=log_dir,
+    ),
+    tf.keras.callbacks.TensorBoard(
+        log_dir=log_dir,
+        histogram_freq=1,
+        update_freq=100,
+    ),
+    tf.keras.callbacks.ModelCheckpoint(
+        filepath=chkpt_dir,
+        monitor="val_loss",
+        mode="min",
+        save_best_only=True,
+        save_weights_only=True,
+    ),
+]
+
+"""
+All that is left to do is run fit()!
+"""
+
+print(train_ds)
+print(val_ds)
+history = contrastive_model.fit(
+    train_ds,
+    epochs=PRE_TRAIN_EPOCHS,
+    steps_per_epoch=PRE_TRAIN_STEPS_PER_EPOCH,
+    validation_data=val_ds,
+    validation_steps=VAL_STEPS_PER_EPOCH,
+    callbacks=callbacks,
+)
+
+
+"""
+## Plotting and Evaluation
+"""
+
+plt.figure(figsize=(15, 4))
+plt.subplot(1, 3, 1)
+plt.plot(history.history["loss"])
+plt.grid()
+plt.title(f"{loss.name} - loss")
+
+plt.subplot(1, 3, 2)
+plt.plot(history.history["proj_std"], label="proj")
+if "pred_std" in history.history:
+    plt.plot(history.history["pred_std"], label="pred")
+plt.grid()
+plt.title(f"{loss.name} - std metrics")
+plt.legend()
+
+plt.subplot(1, 3, 3)
+plt.plot(history.history["binary_accuracy"], label="acc")
+plt.grid()
+plt.title(f"{loss.name} - match metrics")
+plt.legend()
+
+plt.show()
```

Comment:
Do you want to comment on the pre-training results? What the readers should notice and so on? It's definitely quite valuable.

Reply:
Lets see what the renders look like and if its warranted I'll updated!